internal/ecies/ecies.go

```diff
@@ -0,0 +1,136 @@
+// This package implements an ECIES (Elliptic Curve Integrated Encryption
+// Scheme) encryption. It uses ChaCha20Poly1305 for encryption and HKDF with
+// SHA256 for key derivation. The package provides functions to encrypt and
+// decrypt messages using a shared secret derived between two parties using ECDH
+// (Elliptic Curve Diffie-Hellman).
+
+package ecies
+
+import (
+	crand "crypto/rand"
+	"crypto/sha256"
+	"fmt"
+	"io"
+
+	"github.com/btcsuite/btcd/btcec/v2"
+	"golang.org/x/crypto/chacha20poly1305"
+	"golang.org/x/crypto/hkdf"
+)
+
+// EncryptSha256ChaCha20Poly1305 encrypts the given message using
+// ChaCha20Poly1305 with a shared secret (usually derived using ECDH between the
+// sender's ephemeral key and the receiver's public key) that is stretched using
+// HKDF with SHA256. The cipher also authenticates the additional data.
+// The output is a byte slice containing:
+//
+//	<12 bytes nonce> <... bytes ciphertext>
+func EncryptSha256ChaCha20Poly1305(sharedSecret [32]byte, msg []byte,
+	additionalData []byte) ([]byte, error) {
+
+	// We begin by stretching the shared secret using HKDF with SHA256.
+	stretchedKey, err := HkdfSha256(sharedSecret[:])
+	if err != nil {
+		return nil, fmt.Errorf("cannot derive hkdf key: %w", err)
+	}
+
+	// We can now create a new ChaCha20Poly1305 AEAD cipher using the
+	// stretched key.
+	aead, err := chacha20poly1305.New(stretchedKey[:])
+	if err != nil {
+		return nil, fmt.Errorf("cannot create new chacha20poly1305 "+
+			"cipher: %w", err)
+	}
+
+	// Select a random nonce, and leave capacity for the ciphertext.
+	nonce := make(
```

Comment:
As we're using a [random nonce here](https://en.wikipedia.org/wiki/ChaCha20-Poly1305#:~:text=%5Bedit%5D-,XChaCha20%2DPoly1305%20%E2%80%93%20extended%20nonce%20variant,-%5Bedit%5D), we should use [`NewX`](https://pkg.go.dev/golang.org/x/crypto/chacha20poly1305#NewX), which is meant for cases where a counter-like nonce isn't used. Basically some extra security margin for when nonces are generated randomly.

Reply:
Ah, yes, makes sense. Changed to X.